### PR TITLE
Implement simple random write in AlluxioJniFuse

### PR DESCRIPTION
**Pull request by ssz1997**
A very simple implementation of random write in AlluxioJniFuse.

Use `RandomAccessFile` to write into a tmp file with `seek()` and `write()`. Upon finish writing and `release()` is called, create a `FileInputStream` to read from the tmp file, then write into Alluxio with `FileOutStream`, 64KB at a time (use the default value of `alluxio.user.local.writer.chunk.size.bytes`).

The script provided in #14708 could run successfully and have expected outcome.

### CODE REVIEW

**Comment by LuQQiu** [comment]
@ssz1997 can you add the new functionality as an add-on function?
if tmp.dir is enabled, go through the new local-then-alluxio path,
if not, go through the original path?

**Comment by LuQQiu** [comment]
Don't modify the default behavior when tmp.dir is not provided

**Comment by ssz1997** [comment]
Good idea. Will do.
Another question @LuQQiu If the written file already exists, what should be the behavior?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 1]
```diff
@@ -567,28 +573,50 @@ private int writeInternal(String path, ByteBuffer buf, long size, long offset, F
...
+        final byte[] dest = new byte[sz];
+        buf.get(dest, 0, sz);
+        tmpFile.seek(offset);
+        tmpFile.write(dest);
+        tmpFile.close();
+      } catch (IOException e) {
+        LOG.error("IOException while writing to {}.", path, e);
+        return -ErrorCodes.EIO();
+      }
+    } else {
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:598]: No need to use else, it makes the code more difficult to read

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 2]
```diff
@@ -567,28 +573,50 @@ private int writeInternal(String path, ByteBuffer buf, long size, long offset, F
...
-      os.write(dest);
-    } catch (IOException e) {
-      LOG.error("IOException while writing to {}.", path, e);
-      return -ErrorCodes.EIO();
+    if (!mTmpFolder.isEmpty()) {
+      RandomAccessFile tmpFile;
+      try {
+        File tmpFolder = new File(mTmpFolder);
+        if (!tmpFolder.exists()) {
+          tmpFolder.mkdirs();
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:581]: mkdirs may have concurrency issues if multiple open(), create() are called, i would prefer to just error out if tmpFolder does not exist and in the tmp dir description, saying that the folder should exist

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 3]
```diff
@@ -5346,6 +5346,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.IGNORE)
           .setScope(Scope.CLIENT)
           .build();
+  public static final PropertyKey FUSE_TMP_FOLDER =
+      new Builder(Name.FUSE_TMP_FOLDER)
+          .setDefaultValue("")
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:5351]: we can remove this line, i remember there is a Configuration.isSet function to checking whether a variable is set

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 4]
```diff
@@ -5346,6 +5346,14 @@ public String toString() {
...
+          .setDescription("A temporary folder for writing files through fuse. Files are first "
+              + "written to this folder, then written to Alluxio upon finish writing the file")
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:5353]: Add a description about local-then-alluxio write supports random write.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 5]
```diff
@@ -633,9 +661,24 @@ public int release(String path, FuseFileInfo fi) {
   private int releaseInternal(String path, FuseFileInfo fi) {
     long fd = fi.fh.get();
     try {
+      CreateFileEntry<FileOutStream> ce = mCreateFileEntries.getFirstByField(ID_INDEX, fd);
+      if (!mTmpFolder.isEmpty() && ce != null) {
+        FileOutStream os = ce.getOut();
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:666]: No need to create FileOutStream until release
we only need Alluxio FileOutStream in release, separating them is more error-prone

**Review by LuQQiu — COMMENTED** [review]
Left some comments, thanks

**Comment by maobaolong** [comment]
@ssz1997 Before deep review, can you explain how to deal fuse crash and restart and some other error situation.

**Comment by ssz1997** [comment]
@maobaolong Thanks for paying attention to this. 
First of all, this is a very simple implementation of the random write. The purpose for now is not to develop a comprehensive feature, and as you could imagine, random write would require a lot of discussions. That's being said, we will try to make it handle some of the cases/corner cases/failures, but still this is not a comprehensive implementation of random write. @LuQQiu Do we need Baolong review this at all?
Second, to answer your question. Since this is the super simple version, we are mostly considering the perfect world, for now. That is to say, we will consider all the failure cases when we develop the comprehensive random write, but not now. Although if the written file (or its corresponding tmp file) already exists, we will error out.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 6]
```diff
@@ -5346,6 +5346,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.IGNORE)
           .setScope(Scope.CLIENT)
           .build();
+  public static final PropertyKey FUSE_TMP_FOLDER =
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:5349]: prefer `dir` to `folder`
```suggestion
  public static final PropertyKey FUSE_TMP_DIR=
```

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 3 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:5351]: prefer not depending on `Configuration.isSet` to ease maintaining

**Comment by ssz1997** [comment]
@maobaolong Thanks for paying attention to this.

This is a first-step, very simple implementation of the random write. The purpose is not to develop a comprehensive feature, and as you could imagine, random write would require a lot of discussions. That's being said, we don't support complicated scenarios, for example, fuse restart. We definitely will consider these in the next step though.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 2 above)
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:581]: why we put the creation of this temp dir here inside the critical path of `writeInternal` method?
this is only needed once and in this way each `writeInternal` call (could be many) will check existence and make it slow.
can we move it to constructor to prepare the temp dir only once if this is enabled?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 7]
```diff
@@ -633,9 +661,24 @@ public int release(String path, FuseFileInfo fi) {
...
+            Paths.get(mTmpFolder, path).toString());
+        byte[] buf = new byte[64 * 1024];
+        int bytesRead = 0;
+        while (bytesRead != -1) {
+          bytesRead = tmpFileInputStream.read(buf);
+          if (bytesRead > 0) {
+            os.write(buf, 0, bytesRead);
+          }
+        }
+        tmpFileInputStream.close();
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:677]: why not using  `ByteStreams.copy` in Guava to copy from out stream to another?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 8]
```diff
@@ -633,9 +661,24 @@ public int release(String path, FuseFileInfo fi) {
   private int releaseInternal(String path, FuseFileInfo fi) {
     long fd = fi.fh.get();
     try {
+      CreateFileEntry<FileOutStream> ce = mCreateFileEntries.getFirstByField(ID_INDEX, fd);
+      if (!mTmpFolder.isEmpty() && ce != null) {
+        FileOutStream os = ce.getOut();
+        FileInputStream tmpFileInputStream = new FileInputStream(
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:667]: try-with-resource is always preferred

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 9]
```diff
@@ -633,9 +661,24 @@ public int release(String path, FuseFileInfo fi) {
...
+        byte[] buf = new byte[64 * 1024];
+        int bytesRead = 0;
+        while (bytesRead != -1) {
+          bytesRead = tmpFileInputStream.read(buf);
+          if (bytesRead > 0) {
+            os.write(buf, 0, bytesRead);
+          }
+        }
+        tmpFileInputStream.close();
+        Files.delete(Paths.get(mTmpFolder, path));
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:678]: shall we suppress IOE thrown here to hide the case when we failed to delete?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 10]
```diff
@@ -567,28 +573,50 @@ private int writeInternal(String path, ByteBuffer buf, long size, long offset, F
...
+      } catch (FileNotFoundException e) {
+        LOG.error("Failed to create temporary file {}: ", Paths.get(mTmpFolder, path), e);
+        return -ErrorCodes.EIO();
+      }
+      try {
+        final byte[] dest = new byte[sz];
+        buf.get(dest, 0, sz);
+        tmpFile.seek(offset);
+        tmpFile.write(dest);
+        tmpFile.close();
```
Comment (apc999) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:593]: if seek or write throws IOE, this tmpFile will not be closed.
that's why we shall use try-with-resource.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 5 above)
Comment (apc999) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:666]: agree

**Review by apc999 — CHANGES_REQUESTED** [review]
a few comments left

**Comment by LuQQiu** [comment]
@ssz1997 can you help solve the `Checkstyle, Findbugs, Doc Check, etc. / build (pull_request) ` build failure

**Comment by maobaolong** [comment]
@bzheng888 Please take a look at this PR, thanks

**Comment by ssz1997** [comment]
Closing due to no longer need this

**Comment by maobaolong** [comment]
@ssz1997 Thanks for your reply. is there any other plan or approach to support random write for now?

**Pull request closed without merging**